[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38502 [pytorch] move tracing logic to a separate dispatch backend (part 2: integration)**
* #38467 [pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)

This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

ljk53 committed:
[pytorch][WIP] move tracing logic to a separate dispatch backend (part 2: integration)



Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][WIP] move tracing logic to a separate dispatch backend (part 2: integration)"






Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][WIP] move tracing logic to a separate dispatch backend (part 2: integration)"






Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -136,6 +137,7 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
       sizes_({(int64_t)values.size()}), \
       scalar_type_(at::k##S), \
       type_(TensorDataContainerType::Tensor) { \
+    at::tracer::impl::NoTracerDispatchMode tracer_guard; \
```
I'm not entirely sure whether these methods in this header file can be called in the context of tracing. @yf225 - could you please help review this part? Please check the NOTE above for the context.

yf225 replied:
I believe the methods in this header file are strictly C++ frontend code and won't be called by JIT tracing, and it should be okay to omit this guard if we want :)

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -288,6 +289,7 @@ Tensor internal_new_from_data(
   // here.
   Tensor tensor;
   {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
```
Per the comment above, `AutoNonVariableTypeMode` can be deleted here

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -837,6 +838,7 @@ at::Tensor PythonArgs::tensor_slow(int i) {
         Py_TYPE(obj)->tp_name);
   }
+  at::tracer::impl::NoTracerDispatchMode tracer_guard;
   at::AutoNonVariableTypeMode guard;
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -79,6 +80,7 @@ static inline Variable valueToTensor(c10::TensorOptions options, PyObject* value
     return reinterpret_cast<THPVariable*>(value)->cdata;
   }
+  at::tracer::impl::NoTracerDispatchMode tracer_guard;
   at::AutoNonVariableTypeMode guard;
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -273,6 +276,7 @@ Tensor& resize_as_(
 #endif
   {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -251,6 +253,7 @@ Tensor& resize_(
 #endif
   {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -222,6 +223,7 @@ Tensor & copy_(Tensor & self, const Tensor & src, bool non_blocking) {
   }
   {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -222,6 +223,7 @@ Tensor & copy_(Tensor & self, const Tensor & src, bool non_blocking) {
     grad_fn->src_device = src.device();
   }
   {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
```
The model here is something like TracingTypeManual

ljk53 replied:
agreed - I'll add a TODO here. The reason I added the guard is because these manual tracing logic doesn't have `setTracingState(null)` compared to auto-generated TraceType, I should either add `setTracingState(null)` above or use the guard, otherwise it might break old behavior.

ezyang commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -219,6 +222,7 @@ AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TENSOR)
       // `N` is the number of the elements in the tensor).
       at::Tensor tensor = ([&]() {
+        at::tracer::impl::NoTracerDispatchMode tracer_guard;
         at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -209,6 +211,7 @@ AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16, TENSOR)
 
     if (is_scalar()) {
+      at::tracer::impl::NoTracerDispatchMode tracer_guard;
       at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode. I infer that these are tracing only modifications because there is an un-guarded `to` call in the else branch below

ezyang commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -110,6 +114,7 @@ inline at::Tensor from_blob(
     const at::TensorOptions& options = at::TensorOptions()) {
   at::Tensor tensor = ([&]() {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -95,6 +98,7 @@ inline at::Tensor from_blob(
     const at::TensorOptions& options = at::TensorOptions()) {
   at::Tensor tensor = ([&]() {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -77,6 +79,7 @@ inline at::Tensor from_blob(
     const at::TensorOptions& options = at::TensorOptions()) {
   at::Tensor tensor = ([&]() {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -60,6 +61,7 @@ inline at::Tensor from_blob(
     const at::TensorOptions& options = at::TensorOptions()) {
   at::Tensor tensor = ([&]() {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

ezyang commented on c10/core/DispatchKey.cpp:
```diff
@@ -48,6 +48,8 @@ const char* toString(DispatchKey t) {
     case DispatchKey::Profiler:
       return "Profile";
+    case DispatchKey::Tracer:
+      return "Trace";
```
This should be Tracer (exposing the bug above, where it should be Profiler, not Profile)

ezyang commented on aten/src/ATen/cpp_custom_type_hack.h:
```diff
@@ -34,6 +35,7 @@ T& cast(const Tensor& packed) {
 Tensor create(std::unique_ptr<T> ptr, TensorOptions options) {
   // None of this should trace, so turn off Variable handling
+  at::tracer::impl::NoTracerDispatchMode tracer_guard;
   at::AutoNonVariableTypeMode guard;
```
Delete AutoNonVariableTypeMode

ezyang commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -461,6 +462,7 @@ Tensor scalar_tensor(Scalar s, const TensorOptions& options) {
     // revert this to following:
     //   auto result = at::empty({}, options);
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
     at::AutoNonVariableTypeMode non_var_type_mode(true);
```
Delete AutoNonVariableTypeMode

Actually, TBH, I don't think the `NoTracerDispatchMode` is doing anything here either, because neither `empty_cpu` nor `fill_` should be hitting device dispatch

ljk53 replied:
Why did we introduce `AutoNonVariableTypeMode` here in the first place?
I agree empty_cpu / fill_ seems should be simple leaf functions, but I wasn't quite certain how reliable this assumption is (to be future-proof)

ezyang commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -28,6 +29,7 @@ inline Tensor from_blob(
     const TensorOptions& options = {},
     const c10::optional<Device> target_device = c10::nullopt) {
+  tracer::impl::NoTracerDispatchMode tracer_guard;
   AutoNonVariableTypeMode guard;
```
Delete AutoNonVariableTypeMode

ezyang commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -59,6 +61,7 @@ inline Tensor from_blob(
     IntArrayRef strides,
     const TensorOptions& options = {}) {
+  tracer::impl::NoTracerDispatchMode tracer_guard;
   AutoNonVariableTypeMode guard;
```
Delete AutoNonVariableTypeMode

ezyang commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -136,6 +137,7 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
       scalar_type_(at::k##S), \
       type_(TensorDataContainerType::Tensor) { \
+    at::tracer::impl::NoTracerDispatchMode tracer_guard; \
     at::AutoNonVariableTypeMode non_var_type_mode(true); \
```
Delete AutoNonVariableTypeMode

ljk53 replied:
I'll revert changed to this file according to @yf225 's comment.

@yf225 - do you know why AutoNonVariableTypeMode is needed here?

ezyang commented:
This all broadly looks pretty good, and the green CI is also a good indicator. How does the situation look inside fbcode?

ljk53 commented:
> This all broadly looks pretty good, and the green CI is also a good indicator. How does the situation look inside fbcode?

Internal ones all passed too (except a few "no-signal" ones). I'll delete the `AutoNonVariableTypeMode` instances you pointed out and merge this PR into the previous one.

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 2: integration)"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key `Tracer`.

It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline `[Tracing Mode Switches]` note.

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21579367/)!

Differential Revision: [D21579367](https://our.internmc.facebook.com/intern/diff/D21579367)

[ghstack-poisoned]

ljk53 commented:
Addressed most comments and merged into #38467.
